### ISSUE TITLE
Simulation of large core count system

### ISSUE THREAD

**Issue by vijay4454** [s5z/zsim#97]
Hello

I am simulating a multi-threaded workload that launches about 140 threads on a large heterogeneous core count system. I have configured 1 beefy(OOO) and 40 wimpy(IPC-1) cores. I have also tested 1 beefy + 80 wimpy, 1 beefy + 10 wimpy, and 1 beefy + 127 wimpy, 4 beefy + 0 wimpy, 40 beefy +0 wimpy systems. I notice the simulation finishes just fine for small core count system, regardless of whether it is a homogeneous or heterogenous systems (based on the small sample set of tested core configurations). But the deadlock detection fires for large core count system, and the simulation exits before completing execution of entire workload.

Does anyone have an idea of what might be going on and why? Will this be fixed if I simply increase the stall cycles after which deadlock detection fires (currently 120 seconds)? How do i know how much should I increase it to? i have currently switched off the deadlock detection and simulation has been going on for more than 45 minutes now, for the 1 beefy + 40 wimpy configuration.

**UPDATE**

I did some more testing. I switched off the deadlock detection. On beefy4 configuration (4 OOO cores), it takes about 1 min for simulation to complete. On beefy 1 + wimpy 80 configuration (total 81 cores), it takes almost 6 hours for simulation to complete. I do have synchronization between the master thread and each of the other threads. The max number of threads active at many point in the workload is (number of simulated cores - 2 = 79). I don't understand why increasing the thread count/simulated core count increases the simulation time. The total data set size of the workload is same for all configurations. There is no synchronization among the threads created using pthread_create. All synchronization is between master thread and each pthread_create thread.

**Comment by dzhang50** [comment]
I don't know the answer to your question, but IPC1 cores are not actually wimpy. Depending on the benchmark, they can perform better than OOO cores.

**Comment by vijay4454** [comment]
I solved this issue. I was using pthread_cancel in my workload source code to cancel all child threads from the main thread. Instead, I removed this pthread_cancel and instead used a pthread_exit in the child threads. It looks like zsim is substantially slowed down when using the pthread_cancel (600x) while the real hardware is only slowed down by 30% when using pthread_cancel instead of pthread_exit. I am not sure why that is. Probably because zsim has an unoptimized emulation of the OS system calls used by pthread_cancel? 

I needed pthread_cancel/pthread_exit because exiting/canceling a thread was dependent on a certain condition to be true.
